app/src/behaviors/behavior_tri_state.c

```diff
@@ -0,0 +1,301 @@
+        return;
+    }
+    LOG_DBG("Tri-state deactivated due to timer");
+    tri_state->is_active = false;
+    trigger_end_behavior(tri_state);
+}
+
+static void clear_tri_state(struct active_tri_state *tri_state) { tri_state->is_active = false; }
+
+struct active_tri_state active_tri_states[ZMK_BHV_MAX_ACTIVE_TRI_STATES] = {};
+
+static struct active_tri_state *find_tri_state(uint32_t position) {
+    for (int i = 0; i < ZMK_BHV_MAX_ACTIVE_TRI_STATES; i++) {
+        if (active_tri_states[i].position == position && active_tri_states[i].is_active) {
+            return &active_tri_states[i];
+        }
+    }
+    return NULL;
+}
+
+static int new_tri_state(uint32_t position, const struct behavior_tri_state_config *config,
+                         struct active_tri_state **tri_state) {
+    for (int i = 0; i < ZMK_BHV_MAX_ACTIVE_TRI_STATES; i++) {
+        struct active_tri_state *const ref_tri_state = &active_tri_states[i];
+        if (!ref_tri_state->is_active) {
+            ref_tri_state->position = position;
+            ref_tri_state->config = config;
+            ref_tri_state->is_active = true;
+            ref_tri_state->is_pressed = false;
+            ref_tri_state->first_press = true;
+            *tri_state = ref_tri_state;
+            return 0;
+        }
+    }
+    return -ENOMEM;
+}
+
+static bool is_other_key_ignored(struct active_tri_state *tri_state, int32_t position) {
+    for (int i = 0; i < tri_state->config->ignored_key_positions_len; i++) {
+        if (tri_state->config->ignored_key_positions[i] == position) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool is_layer_ignored(struct active_tri_state *tri_state, int32_t layer) {
+    if ((BIT(layer) & tri_state->config->ignored_layers) != 0U) {
+        return true;
+    }
+    return false;
+}
+
+static int on_tri_state_binding_pressed(struct zmk_behavior_binding *binding,
+                                        struct zmk_behavior_binding_event event) {
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_tri_state_config *cfg = dev->config;
+    struct active_tri_state *tri_state;
+    tri_state = find_tri_state(event.position);
+    if (tri_state == NULL) {
+        if (new_tri_state(event.position, cfg, &tri_state) == -ENOMEM) {
+            LOG_ERR("Unable to create new tri_state. Insufficient space in "
+                    "active_tri_states[].");
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+        LOG_DBG("%d created new tri_state", event.position);
+    }
+    LOG_DBG("%d tri_state pressed", event.position);
+    tri_state->is_pressed = true;
+    if (tri_state->first_press) {
+        behavior_keymap_binding_pressed((struct zmk_behavior_binding *)&cfg->start_behavior, event);
+        behavior_keymap_binding_released((struct zmk_behavior_binding *)&cfg->start_behavior,
+                                         event);
```

Comment:
I know we do this for hold taps right now, but any reason not to use the behavior queue here?